rpc_util.go

```diff
@@ -289,53 +289,63 @@ func (p *parser) recvMsg(maxReceiveMessageSize int) (pf payloadFormat, msg []byt
 
 // encode serializes msg and prepends the message header. If msg is nil, it
 // generates the message header of 0 message length.
-func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayload *stats.OutPayload) ([]byte, error) {
+func encode(c Codec, msg interface{}, cp Compressor, outPayload *stats.OutPayload) ([]byte, error) {
 	var (
-		b      []byte
-		length uint
+		b          []byte
+		length     uint
+		payloadLen = 1
+		sizeLen    = 4
 	)
-	if msg != nil {
-		var err error
-		// TODO(zhaoq): optimize to reduce memory alloc and copying.
-		b, err = c.Marshal(msg)
-		if err != nil {
-			return nil, Errorf(codes.Internal, "grpc: error while marshaling: %v", err.Error())
-		}
-		if outPayload != nil {
-			outPayload.Payload = msg
-			// TODO truncate large payload.
-			outPayload.Data = b
-			outPayload.Length = len(b)
-		}
-		if cp != nil {
-			if err := cp.Do(cbuf, b); err != nil {
-				return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
-			}
-			b = cbuf.Bytes()
+	if msg == nil {
+		var buf = make([]byte, payloadLen+sizeLen)
+		return buf, nil
+	}
+
+	var err error
+	// TODO(zhaoq): optimize to reduce memory alloc and copying.
+	b, err = c.Marshal(msg)
+	if err != nil {
+		return nil, Errorf(codes.Internal, "grpc: error while marshaling: %v", err.Error())
+	}
+	if outPayload != nil {
+		outPayload.Payload = msg
+		// TODO truncate large payload.
+		outPayload.Data = b
+		outPayload.Length = len(b)
+	}
+
+	var cbuf *bytes.Buffer
+
+	if cp != nil {
+		// pre-alloc place
+		cbuf = bytes.NewBuffer(make([]byte, payloadLen+sizeLen))
+		if err := cp.Do(cbuf, b); err != nil {
+			return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
 		}
+		b = cbuf.Bytes()
 		length = uint(len(b))
 	}
+
+	length = uint(len(b))
 	if length > math.MaxUint32 {
 		return nil, Errorf(codes.ResourceExhausted, "grpc: message too large (%d bytes)", length)
 	}
-
-	const (
-		payloadLen = 1
-		sizeLen    = 4
-	)
-
-	var buf = make([]byte, payloadLen+sizeLen+len(b))
-
-	// Write payload format
-	if cp == nil {
-		buf[0] = byte(compressionNone)
-	} else {
+	var buf []byte
+	if cp != nil {
+		// re-use b
+		buf = b
+		// Write payload format
 		buf[0] = byte(compressionMade)
+		// Write length of b into buf
+		binary.BigEndian.PutUint32(buf[1:], uint32(length-5))
```

Comment:
With length set correctly, you can share this code.  the only different parts with cp vs. without is that without, you need to allocate another buffer and copy the message.

Actually, can encode() just *not* do this part, and we leave it up to wherever we actually write the data to the wire, to avoid the extra copy and allocation?  It's a little gross, because we'll have to have a "if cp != nil" check elsewhere for "compressionMade", but it would be worth it to avoid a potentially big allocation.

So, essentially, encode() returns the encoded message itself only.